diff --git a/node_modules/@auto-it/core/dist/auto.js b/node_modules/@auto-it/core/dist/auto.js
index b388374..41949be 100644
--- a/node_modules/@auto-it/core/dist/auto.js
+++ b/node_modules/@auto-it/core/dist/auto.js
@@ -316,7 +316,7 @@ class Auto {
                 : undefined,
         };
         const envVar = Object.keys(GIT_TOKENS).find((v) => process.env[v]) || "";
-        const gitCredentials = GIT_TOKENS[envVar] || process.env.GH_TOKEN;
+        const gitCredentials = process.env.GH_TOKEN || GIT_TOKENS[envVar];
         if (ssh_url && (await this.git.verifyAuth(ssh_url))) {
             this.logger.veryVerbose.note("Using ssh URL as remote");
             return ssh_url;
